src/pd/async/util.rs

```diff
@@ -0,0 +1,301 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::sync::TryLockError;
+use std::time::Instant;
+use std::time::Duration;
+use std::thread;
+
+use futures::Poll;
+use futures::Async;
+use futures::Future;
+use futures::future::ok;
+use futures::future::{self, loop_fn, Loop};
+
+use kvproto::pdpb::GetMembersResponse;
+use kvproto::pdpb_grpc::PDAsync;
+use kvproto::pdpb_grpc::PDAsyncClient;
+
+use util::HandyRwLock;
+
+use super::super::PdFuture;
+use super::super::Result;
+use super::super::Error;
+use super::client::try_connect_leader;
+
+#[derive(Debug)]
+struct Bundle<C, M> {
```

Comment:
need to use template here? maybe the type C is PDAsyncClient and M is GetMembersResponse all the time.